### DISCUSSION TITLE
Extra prompt on exit

### DISCUSSION THREAD

**Discussion by grbba** [daniele77/cli#103] [Q&A]
Sorry for asking again a maybe stupid question:

on exit i get an extra prompt with an additional character
```
DccEX> exit
Closing App...
Goodbye and thanks for all the fish.
DccEX> %
```     
any idea where that may come from ?
//

**Comment by daniele77** [comment]
Do you face the same problem with the `complete` example?

I think the problem could be due to your mechanism of shutdown: if the local session exit action is a `scheduler.Stop()` and you have only one thread in the application, you shouldn't see any extra prompt or character on exit.

    CliLocalTerminalSession localSession(cli, scheduler, std::cout, 200);
    localSession.ExitAction(
        [&scheduler](auto& out) // session exit action
        {
            out << "Closing App...\n";
            scheduler.Stop();
        }
    );

**Reply by grbba** [reply]
yes happens also on the complete example :
```
 - echo <first string to echo> <second string to echo>
        Print the strings passed as parameter
 - error
        Throw an exception in the command handler
 - reverse <string_to_revert>
        Print the reverse string
 - add <first_term> <second_term>
        Print the sum of the two numbers
 - add <int> <int> <int>
        Print the sum of the three numbers
 - color
        Enable colors in the cli
 - nocolor
        Disable colors in the cli
 - removecmds

 - sub
        (menu)
cli> exit
Closing App...
Goodbye and thanks for all the fish.
cli> %       
```                     

It happens actually with all the examples except **simplelocalsession** and  **filesession** (run with an empty input.txt)  where in the output.txt only this line `cli> Goodbye and thanks for all the fish.` appears not the extra prompt nor the %

**Reply by grbba** [reply]
Ok getting closer i am running zsh ( default on Mac these days ) not bash or sh
for bash it looks like this i.e. the last % doesn't appear

```
grbba % bash
bash-5.1$ ./complete 
cli> exit
Closing App...
Goodbye and thanks for all the fish.
cli> bash-5.1$ 
```
same for sh
```
grbba % sh
sh-3.2$ ./complete
cli> exit
Closing App...
Goodbye and thanks for all the fish.
cli> sh-3.2$ 
```
for zsh
```
grbba % ./complete 
cli> exit
Closing App...
Goodbye and thanks for all the fish.
cli> %                                                                                                                                                                                                                                                 
grbba %
```

**Comment by daniele77** [comment]
On my bash:

    daniele@daniele-kubuntu-18:~/projects/cli/build/examples$ ./complete 
    cli> color 
    Colors ON
    cli> exit
    Closing App...
    Goodbye and thanks for all the fish.
    cli> daniele@daniele-kubuntu-18:~/projects/cli/build/examples$ 

You always get the shell prompt (in my case: `daniele@daniele-kubuntu-18:~/projects/cli/build/examples$`) after the cli prompt (i.e., `cli> `).
I think this can't be considered an issue.

I don't have (yet) zsh installed to test it, but from your comment, it seems you get:

    cli> %                                                                                                                                                                                                                                                 
    grbba %

after exit. I.e.,:
1. a cli prompt (`cli> `)
2. a zsh % + a newline
3. a zsh prompt (`grbba %`)

Can this behavior be considered an issue for your application?

**Reply by grbba** [reply]
Hello, sorry for the late reply. No it's not an issue but a bit confusing that you get an extra prompt after exiting the app. I know I will get questions about this :) but it's manageable.
I have also solved all the other outstanding issues, so far, with the async serial communications i had to set VMIN and VTIME attributes properly on the port and now it all works including exiting properly :). They were set by default to block on the read permanently until at least 1 char arrives so i never got the chance to actually close the port properly on exit. 

the only thing i can't really solve is the fact that the promt will show up possibly in the middle of the output coming back from the MCU as its async,  and i don't know how much will come back at any given time and how long it will take ... I added a short delay in the lamdas of the commands to allow for the reply from the MCU so that the prompt will show up in 98% of all times after the feedback from the MCU.

Have a nice day
//

**Reply by grbba** [reply]
I just had an idea and it works .. maybe not the best 

```
 cli.ExitAction(
      [&](auto &out) {
        serial.closePort();
        out << "Goodbye and thanks for all the steam.\n";
        std::cout.setstate(std::ios_base::badbit);      // prevents any additional output just before exiting
      });
```

I added the badbit setting to std out in the ExitAction and the prompt is gone :)

**Reply by daniele77** [reply]
I'll try to find a solution in the library, ASAP
